grpc_stdio.go

```diff
@@ -0,0 +1,196 @@
+// the stdio data to output writers.
+type grpcStdioClient struct {
+	log         hclog.Logger
+	stdioClient plugin.GRPCStdio_StreamStdioClient
+}
+
+// newGRPCStdioClient creates a grpcStdioClient. This will perform the
+// initial connection to the stdio service. If the stdio service is unavailable
+// then this will be a no-op. This allows this to work without error for
+// plugins that don't support this.
+func newGRPCStdioClient(
+	ctx context.Context,
+	log hclog.Logger,
+	conn *grpc.ClientConn,
+) (*grpcStdioClient, error) {
+	client := plugin.NewGRPCStdioClient(conn)
+
+	// Connect immediately to the endpoint
+	stdioClient, err := client.StreamStdio(ctx, &empty.Empty{})
+
+	// If we get an Unavaiable error, this means that the plugin isn't
+	// updated and linking to the latest version of go-plugin that supports
+	// this. We fall back to the previous behavior of just not syncing anything.
+	if status.Code(err) == codes.Unavailable {
+		log.Warn("stdio service not available, stdout/stderr syncing unavailable")
+		stdioClient = nil
+		err = nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return &grpcStdioClient{
+		log:         log,
+		stdioClient: stdioClient,
+	}, nil
+}
+
+// Run starts the loop that receives stdio data and writes it to the given
+// writers. This blocks and should be run in a goroutine.
+func (c *grpcStdioClient) Run(stdout, stderr io.Writer) {
+	// This will be nil if stdio is not supported by the plugin
+	if c.stdioClient == nil {
+		c.log.Warn("stdio service unavailable, run will do nothing")
+		return
+	}
+
+	for {
+		c.log.Trace("waiting for stdio data")
+		data, err := c.stdioClient.Recv()
+		if err != nil {
+			if err == io.EOF ||
+				status.Code(err) == codes.Unavailable ||
+				err == context.Canceled {
+				c.log.Warn("received EOF, stopping recv loop", "err", err)
+				return
+			}
+
+			c.log.Error("error receiving data", "err", err)
+			continue
+		}
+
+		// Determine our output writer based on channel
+		var w io.Writer
+		switch data.Channel {
+		case plugin.StdioData_STDOUT:
+			w = stdout
+
+		case plugin.StdioData_STDERR:
+			w = stderr
+
+		default:
+			c.log.Warn("unknown channel, dropping", "channel", data.Channel)
+			continue
+		}
+
+		// Write! In the event of an error we just continue.
+		c.log.Trace("received data", "channel", data.Channel.String(), "len", len(data.Data))
+		if _, err := io.Copy(w, bytes.NewReader(data.Data)); err != nil {
+			c.log.Error("failed to copy all bytes", "err", err)
+		}
+	}
+}
+
+// copyChan copies an io.Reader into a channel.
+func copyChan(dst chan<- []byte, src io.Reader) {
+	bufsrc := bufio.NewReader(src)
+
+	for {
+		// Make our data buffer. We allocate a new one per loop iteration
+		// so that we can send it over the channel.
+		var data [1024]byte
+
+		// Read the data, this will block until data is available
+		n, err := bufsrc.Read(data[:])
+
+		// If we hit EOF we're done copying
+		if err == io.EOF {
```

Comment:
Data should be sent before this check

Reply:
The stdlib docs state that there will never be any data on `io.EOF` so we don't need to send anything (`n == 0` guaranteed). Am I misunderstanding?

Reply:
Note I did update to send data before checking `err != nil` AFTER this though, since the above guarantee is only made on EOF. I guess it doesn't hurt to do it anytime but let me know what you think.

Reply:
Since this is a generic `io.Reader`, we should follow the broadest form of the spec. I actually can't recall which concrete types behave which way, since we always need to cover them all in read loops like this. The portion of the stdlib that documents this is in `io.Reader` (with emphasis on why I always check this in review): 

> When Read encounters an error or end-of-file condition after successfully reading n > 0 bytes, it returns the number of bytes read. _It may return the (non-nil) error from the same call_ or return the error (and n == 0) from a subsequent call. An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF.
> 
> _Callers should always process the n > 0 bytes returned before considering the error err_. Doing so correctly handles I/O errors that happen after reading some bytes and also both of the allowed EOF behaviors.

Reply:
Sounds good. It isn't a generic reader though since I always wrap it in `bufio.Reader`. But I'll just check the bytes earlier, no harm no foul!